// @hexlet/immutable-fs-trees/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filter = exports.reduce = exports.map = exports.isDirectory = exports.isFile = exports.getName = exports.getMeta = exports.getChildren = exports.mkdir = exports.mkfile = void 0;

// @ts-check

/**
 * Make file node
 * @example
 * mkfile('config.json');
 * // {
 * //   name: 'config.json',
 * //   meta: {},
 * //   type: 'file',
 * // }
 *
 * mkfile('config.json', { size: 1200 });
 * // {
 * //   name: 'config.json',
 * //   meta: { size: 1200 },
 * //   type: 'file',
 * // }
 */
const mkfile = (name, meta = {}) => ({
  name,
  meta,
  type: 'file'
});
/**
 * Make directory node
 * @example
 * mkdir('etc');
 * // {
 * //   name: 'etc',
 * //   children: [],
 * //   meta: {},
 * //   type: 'directory',
 * // }
 *
 * mkdir('etc', [mkfile('config'), mkfile('hosts')], { owner: 'user' });
 * // {
 * //   name: 'etc',
 * //   children: [
 * //     { name: 'config', meta: {}, type: 'file' },
 * //     { name: 'hosts', meta: {}, type: 'file' }
 * //   ],
 * //   meta: { owner: 'user' },
 * //   type: 'directory',
 * // }
 */


exports.mkfile = mkfile;

const mkdir = (name, children = [], meta = {}) => ({
  name,
  children,
  meta,
  type: 'directory'
});
/**
 * Return children
 * @example
 * getChildren(mkdir('etc')); // []
 * getChildren(mkdir('etc', [mkfile('name')])); // [<file>]
 */


exports.mkdir = mkdir;

const getChildren = directory => directory.children;
/**
 * Return meta
 * @example
 * getMeta(mkfile('etc')); // {}
 * getMeta(mkfile('etc', { owner: 'root' })); // { owner: 'root' }
 */


exports.getChildren = getChildren;

const getMeta = node => node.meta;
/**
 * Return name
 * @example
 * getName(mkfile('etc')); // etc
 * getName(mkdir('/')); // /
 */


exports.getMeta = getMeta;

const getName = node => node.name;
/**
 * Check is node a file
 * @example
 * isFile(mkfile('config')); // true
 * isFile(mkdir('etc')); // false
 */


exports.getName = getName;

const isFile = node => node.type === 'file';
/**
 * Check is node a directory
 * @example
 * isDirectory(mkdir('etc')); // true
 * isDirectory(mkfile('config')); // false
 */


exports.isFile = isFile;

const isDirectory = node => node.type === 'directory';
/**
 * Map tree
 * @example
 * const tree = mkdir('etc', [mkfile('config'), mkfile('hosts')]);
 *
 * const callbackFn = (node) => {
 *   const { name } = node;
 *   const newName = name.toUpperCase();
 *   return { ...node, name: newName };
 * };
 *
 * map(callbackFn, tree);
 * // {
 * //   name: 'ETC',
 * //   children: [
 * //     { name: 'CONFIG', meta: {}, type: 'file' },
 * //     { name: 'HOSTS', meta: {}, type: 'file' }
 * //   ],
 * //   meta: {},
 * //   type: 'directory',
 * // }
 */


exports.isDirectory = isDirectory;

const map = (callbackFn, tree) => {
  const updatedNode = callbackFn(tree);
  return isDirectory(tree) ? { ...updatedNode,
    children: tree.children.map(n => map(callbackFn, n))
  } : updatedNode;
};
/**
 * Reduce tree
 * @example
 * const tree = mkdir('etc', [mkfile('config'), mkfile('hosts')]);
 *
 * reduce((acc) => acc + 1, tree, 0);
 * // 3
 *
 * reduce((acc, node) => [...acc, node.name], tree, []);
 * // ['etc', 'config', 'hosts']
 */


exports.map = map;

const reduce = (callbackFn, tree, acc) => {
  const newAcc = callbackFn(acc, tree);

  if (isFile(tree)) {
    return newAcc;
  }

  return tree.children.reduce((iAcc, n) => reduce(callbackFn, n, iAcc), newAcc);
};
/**
 * Filter tree
 * @example
 * const tree = mkdir('etc', [mkfile('CONFIG'), mkfile('hosts')]);
 *
 * const callbackFn = (node) => {
 *   const { name } = node;
 *   return name === name.toLowerCase();
 * };
 *
 * filter(callbackFn, tree);
 * // {
 * //   name: 'etc',
 * //   children: [
 * //     { name: 'hosts', meta: {}, type: 'file' }
 * //   ],
 * //   meta: {},
 * //   type: 'directory',
 * // }
 */


exports.reduce = reduce;

const filter = (callbackFn, tree) => {
  if (!callbackFn(tree)) {
    return null;
  }

  return isDirectory(tree) ? { ...tree,
    children: tree.children.map(n => filter(callbackFn, n)).filter(v => v)
  } : tree;
};

exports.filter = filter;
